### PR TITLE
CNTRLPLANE-3167: support STS/IRSA credentials and standalone Velero

### PR DESCRIPTION
**Pull request by jparrill**
## Summary

- Add AWS STS/IRSA credential auto-detection and pre-signing support for etcd snapshot restore (pure stdlib STS client, no AWS SDK dependency)
- Detect OADP+DPA vs standalone Velero at plugin initialization via DPA CRD check
- Fall back to well-known `cloud-credentials` secret (key: `cloud`) when BSL has no `.spec.credential` reference (ARO with Workload Identity, future ROSA without DPA)
- Preserve original secret key alongside `credentials` in copied secret so the HCPEtcdBackup controller can auto-detect credential type

## Description

ROSA HCP and ARO HCP use short-lived federated credentials instead of long-lived static keys. This PR enables the OADP plugin to handle these environments:

- **AWS STS/IRSA**: `ParseAWSCredentialData` auto-detects 3 credential formats (bare ARN, INI with `role_arn`, static keys). When IRSA credentials are detected during restore, the plugin assumes the role via `AssumeRoleWithWebIdentity` before pre-signing the S3 URL
- **Standalone Velero detection**: Plugin init checks for the DPA CRD (`dataprotectionapplications.oadp.openshift.io`). When absent (ARO, future ROSA without DPA), falls back to the well-known `cloud-credentials` secret in the Velero namespace
- **Azure Workload Identity**: No pre-signing needed — the Azure Blob URL passes through directly, and the HCPEtcdBackup controller accesses the blob via projected SA token
- **Secret key preservation**: `copyCredentialSecret` now copies the credential data with both `credentials` (for S3 controller) and the original key (e.g., `cloud` for Azure WI detection)

### Scenarios covered

| Scenario | Backup (copy creds to HO NS) | Restore (inject URL) |
|----------|------------------------------|----------------------|
| ROSA + OADP/DPA | BSL `.spec.credential` (existing flow) | Pre-sign S3 with BSL creds (existing flow) |
| ROSA + Velero (future) | Fallback `cloud-credentials` in velero NS | Pre-sign S3 with fallback + STS |
| ARO + Velero | Fallback `cloud-credentials` in velero NS | No pre-sign — Azure Blob URL passed through, controller accesses via WI |

### Related PRs

- Controller side: [openshift/hypershift#8368](https://github.com/openshift/hypershift/pull/8368) — HCPEtcdBackup controller auto-detect credential mode (STS/IRSA for AWS, Workload Identity for Azure)

## Jira

- [CNTRLPLANE-3167](https://issues.redhat.com/browse/CNTRLPLANE-3167)

## Test plan

- [x] `go build ./...` passes
- [x] All unit tests pass (`pkg/s3presign`, `pkg/core`, `pkg/etcdbackup`)
- [x] Existing tests with BSL `.spec.credential` present pass (no regression)
- [x] New tests: BSL without credential ref — fallback success + fallback with missing secret error
- [x] New tests: `ParseAWSCredentialData` — bare ARN, INI STS, static keys, invalid input
- [x] New tests: `STSClient.AssumeRoleWithWebIdentity` — success, error, XML parsing
- [ ] E2E: ROSA STS cluster — trigger restore, verify plugin assumes role via STS and pre-signs URL
- [ ] E2E: ARO cluster — trigger backup with standalone Velero, verify fallback credential copy works

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Azure Blob presigning with AAD user-delegation SAS and account-key SAS
  * AWS credential type detection including web-identity/role-based flows
  * Automatic fallback to a default credential secret when none is specified
  * Detection of paired vs standalone deployments for credential handling

* **Improvements**
  * Skip etcd StatefulSet restores when appropriate
  * Generate Kubernetes-safe resource names within label length limits
  * More robust cross-cloud credential resolution and signing

* **Tests**
  * Expanded unit coverage for Azure, AWS STS, token providers, and restore flows

### CODE REVIEW

**Comment by tony-schndr** [comment]
@jparrill I have validated image https://quay.io/repository/jparrill/hypershift-oadp-plugin?tab=tags&tag=[CNTRLPLANE-3167](https://redhat.atlassian.net/browse/CNTRLPLANE-3167)-v3 works in ARO-HCP personal development environments by:

1. Creating a Hosted Cluster
2. Creating a test config map on Hosted Cluster
3. Taking an velero backup using `etcdBackupMethod: etcdSnapshot`
4. Deleting the Hosted Cluster
5. Restoring it from the backup from step 3.
6. Validating the test config map is still present on the Hosted Cluster

Exact image used (mirrored from quay.io): arohcpsvcdev.azurecr.io/jparrill/hypershift-oadp-plugin@sha256:8c5c1db7cee57d5612db8952bfffd9ddb44408b2ed1ee773b54638e6e90a9d8f

File: pkg/azblobsas/delegation.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	"encoding/xml"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+var httpDo = http.DefaultClient.Do
```
Comment (mgencur) [inline pkg/azblobsas/delegation.go:16]: Nit: It looks like this httpDo is used only once. Probably makes sense to use `http.DefaultClient.Do` directly and get rid of this variable.

File: pkg/azblobsas/delegation.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+	now := nowFunc().UTC()
+	expiry := opts.Expiry
+	if expiry <= 0 {
+		expiry = DefaultSASExpiry
+	}
+	expiryTime := now.Add(expiry)
+
+	signedStart := now.Format("2006-01-02T15:04:05Z")
+	signedExpiry := expiryTime.Format("2006-01-02T15:04:05Z")
```
Comment (mgencur) [inline pkg/azblobsas/delegation.go:124]: It is strange to see this old date here hardcoded. Does it have any meaning or effect?

File: pkg/azblobsas/delegation.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+		host = parsed.Scheme + "://" + parsed.Host
+	} else {
+		host = fmt.Sprintf("https://%s.blob.core.windows.net", account)
+	}
+
+	reqURL := host + "/?restype=service&comp=userdelegationkey"
+
+	body, err := xml.Marshal(keyInfoRequest{
+		Start:  start.UTC().Format("2006-01-02T15:04:05Z"),
+		Expiry: expiry.UTC().Format("2006-01-02T15:04:05Z"),
```
Comment (mgencur) [inline pkg/azblobsas/delegation.go:54]: It is strange to see this old date here hardcoded. Does it have any meaning or effect? Is it necessary?

File: pkg/core/restore.go [hunk 4]
```diff
@@ -197,28 +213,34 @@ func (p *RestorePlugin) Execute(input *velero.RestoreItemActionExecuteInput) (*v
 		p.log.Debugf("Pod found, skipping restore")
 		return velero.NewRestoreItemActionExecuteOutput(input.Item).WithoutRestore(), nil
 
+	case kind == "StatefulSet":
```
Comment (mgencur) [inline pkg/core/restore.go:216]: For me this doesn't belong to this PR. Is this something that was forgotten in previous PRs? Or, it is related to STS/IRSA. Just curious.

**Review by mgencur — COMMENTED** [review]
Hey @jparrill This looks good! I have a few question here and there. I'm also posting the PR review by `/code-review:pr` from ai-helpers. There are some minor recommendations but overall it's OK:

# Code Review Report: PR #247

## PR Info

- **Title**: CNTRLPLANE-3167: support STS/IRSA credentials and standalone Velero
- **Author**: jparrill (Juan Manuel Parrilla Madrid)
- **Base branch**: main
- **Link**: https://github.com/openshift/hypershift-oadp-plugin/pull/247

## Files Reviewed

**Source files (10)**:

| File | Status |
|------|--------|
| `pkg/azblobsas/delegation.go` | NEW |
| `pkg/azblobsas/sas.go` | NEW |
| `pkg/azblobsas/token.go` | NEW |
| `pkg/common/credentials.go` | NEW |
| `pkg/common/types.go` | MODIFIED |
| `pkg/core/backup.go` | MODIFIED |
| `pkg/core/restore.go` | MODIFIED |
| `pkg/etcdbackup/orchestrator.go` | MODIFIED |
| `pkg/s3presign/presign.go` | MODIFIED |
| `pkg/s3presign/sts.go` | NEW |

**Test files (8)**:

| File | Status |
|------|--------|
| `pkg/azblobsas/delegation_test.go` | NEW |
| `pkg/azblobsas/sas_test.go` | NEW |
| `pkg/azblobsas/token_test.go` | NEW |
| `pkg/core/restore_test.go` | MODIFIED |
| `pkg/etcdbackup/orchestrator_test.go` | MODIFIED |
| `pkg/s3presign/presign_test.go` | MODIFIED |
| `pkg/s3presign/sts_test.go` | NEW |

**Config**: `go.mod`, `go.sum` (dependency updates + new Azure SDK deps)

---

## Build Verification: PASS

| Check | Result |
|-------|--------|
| `go build ./...` | PASS |
| `go vet ./...` | PASS (clean) |
| `go test` (6 packages) | PASS (all ok) |

---

## Unit Test Coverage

**Overall**: Strong. Every new source file has a corresponding `_test.go`. Tests use table-driven patterns with `t.Run()`, cover success and error paths, and test edge cases.

### Coverage gaps (ordered by severity)

| Severity | Location | Gap |
|----------|----------|-----|
| Medium | `pkg/common/credentials.go:11` | `ResolveCredentialRef` has **no direct unit test**. Two branches (explicit ref vs fallback) used in critical backup/restore paths. Tested indirectly but should have a dedicated `credentials_test.go`. |
| Low | `pkg/core/backup.go:266-271` | Etcd PVC exclusion logic (`strings.HasPrefix(name, "data-etcd-")`) is not tested in `backup_test.go`. |
| Low | `pkg/azblobsas/delegation.go:112` | Invalid base64 delegation key error path not tested. |
| Low | `pkg/azblobsas/token.go:27` | `NewAADTokenProvider` not directly tested (acceptable — requires real Azure infra; mock injection in restore tests is correct). |

### Coverage details by file

#### `pkg/azblobsas/delegation.go` (NEW) — Tests: `delegation_test.go`

- `GetUserDelegationKey` — **TESTED** (success, HTTP 403, HTTP 500, malformed XML). Verifies request method, query params, headers, and response parsing.
- `GenerateUserDelegationSASBlobURL` — **TESTED** (basic URL structure, deterministic output, custom endpoint, http endpoint, missing fields, nil delegation key, different blobs produce different sigs, different keys produce different sigs, default expiry).
- Minor gap: no test for invalid endpoint string (`url.Parse` error path at `delegation.go:43`).
- Minor gap: no test for invalid base64 delegation key value (error path at `delegation.go:112`).

#### `pkg/azblobsas/sas.go` (NEW) — Tests: `sas_test.go`

- `ParseAzBlobURL` — **TESTED** (11 cases: valid public/gov/china/http URLs, missing blob, missing container, wrong scheme, non-Azure URL, empty URL, trailing slash).
- `IsAzBlobURL` — **TESTED** (11 cases: all cloud variants, wrong schemes, empty string, non-URL input).
- `ParseAzureCredentials` — **TESTED** (8 cases: standard key, custom env var, comments/blank lines, whitespace, missing key, empty data, value with equals, multiple entries).
- `ParseAzureAADCredentials` — **TESTED** (6 cases: workload identity, service principal, comments, missing tenant, missing client, empty data).
- `GenerateSASBlobURL` — **TESTED** (10 cases: basic structure, determinism, custom endpoint, http endpoint, default expiry, missing fields, missing key, invalid base64, special chars, sig changes).
- No gaps identified.

#### `pkg/azblobsas/token.go` (NEW) — Tests: `token_test.go`

- `NewAADTokenProvider` — **NOT DIRECTLY TESTED**. Acceptable — creates Azure SDK credential objects requiring real infrastructure. Callers inject a mock `TokenProvider`.
- `StorageScopeForCloud` — **TESTED** (6 cases covering all cloud variants and empty string).
- `cloudConfigForName` — **TESTED** (6 cases matching cloud names to authority hosts).

#### `pkg/common/credentials.go` (NEW) — Tests: **NONE**

- `ResolveCredentialRef` — **NOT TESTED** directly. Exercised indirectly through orchestrator and restore tests. A direct unit test is warranted.

#### `pkg/core/restore.go` (MODIFIED) — Tests: `restore_test.go` (MODIFIED)

- `RestorePlugin.Execute` — **TESTED** via `TestRestoreExecuteEtcdStatefulSet`, `TestRestoreExecuteSnapshotURL`, `TestRestoreExecuteHCPSnapshotURL` (etcd StatefulSet skip/restore, HC with s3/https/Azure annotations, HCP with s3/https/Azure annotations, missing BSL).
- `presignS3URL` — **TESTED** (9 cases: valid URL, invalid URL, BSL missing, BSL no cred ref fallback, fallback secret missing, wrong key, https URL, IRSA/STS credentials, STS failure, custom endpoint).
- `presignAzBlobUrl` — **TESTED** via `TestPresignAzBlobUrlWithAAD` (7 cases: AAD with credentials, missing AAD fields, WI env vars, missing env vars, token acquisition failure, account key SAS path).

#### `pkg/etcdbackup/orchestrator.go` (MODIFIED) — Tests: `orchestrator_test.go` (MODIFIED)

- `CreateEtcdBackup` — **TESTED** (7 cases: valid AWS BSL, nil HC, BSL not found, unsupported provider, no credential ref fallback, long name truncation, missing fallback secret).
- `copyCredentialSecret` — **TESTED** (5 cases: key remapping, already exists, key=credentials no duplication, missing key, missing secret).
- `safeResourceName` — **TESTED** (5 cases: short/exact/truncation/trailing hyphen).

#### `pkg/s3presign/presign.go` (MODIFIED) — Tests: `presign_test.go` (MODIFIED)

- `ParseAWSCredentialData` (NEW) — **TESTED** (13 cases: bare ARN, ARN with whitespace, GovCloud/China ARNs, missing env token, INI with role_arn+token, role_arn only env fallback, non-default profile, source_profile error, credential_source error, static creds, static with session token, empty data, whitespace-only, invalid INI).
- `isBareARN` (NEW) — **TESTED** (9 cases).
- `parseSTSProfile` (NEW) — **TESTED** (5 cases).

#### `pkg/s3presign/sts.go` (NEW) — Tests: `sts_test.go` (NEW)

- `STSClient.AssumeRoleWithWebIdentity` — **TESTED** (success, HTTP error, malformed XML, 4 missing-credential-field combinations, missing token file, empty token file, empty roleARN, empty tokenFile path, default session name).
- `NewSTSClient` — **TESTED**.

### Test quality observations

- All test files use table-driven tests with `t.Run()` — consistent with project conventions.
- `t.Helper()` is used appropriately in helper functions (`writeTempFile`, `writeTokenFile`, `newHCPUnstructured`).
- `t.Parallel()` is not used anywhere in the PR. Tests appear independent and could run in parallel, but this is a style choice, not a correctness issue.
- Error messages are verified with substring matching, which is robust against message rewording.
- Mock injection (via `newTokenProvider`, `newSTSClient` function fields) is a clean pattern for testing external service calls.

---

## Idiomatic Go Code

| Severity | Location | Finding | Recommendation |
|----------|----------|---------|----------------|
| **Medium** | `restore.go:339,381` | `presignAzBlobUrl` / `presignAzBlobUrlWithAAD` — "URL" should be all-caps per Go naming convention | Rename to `presignAzBlobURL` / `presignAzBlobURLWithAAD` |
| Low | `restore.go:411` | Uses `time.Now()` directly instead of `nowFunc` — inconsistent with `azblobsas` package's testable time pattern | Use `nowFunc` or accept time as parameter |
| Low | `backup.go:8-9`, `restore.go:13-15` | Redundant import aliases that match package names (`common "...common"`, `validation "...validation"`) | Remove explicit aliases |
| Low | `backup.go:139`, `restore.go:147` | No-op type conversion `context.Context(p.ctx)` | Pre-existing, but should be `p.ctx` directly |
| Info | `presign.go:94,158` | `ParseAWSCredentials` vs `ParseAWSCredentialData` naming overlap could confuse callers | Consider `DetectAWSCredentials` for the higher-level function |
| Info | `presign.go:167` | `os.Getenv` inside a parsing function makes it impure | Consider accepting the env var value as a parameter |
| Info | `orchestrator_test.go:50-52` | Comment/function name mismatch (`TestMapBSLToStorage` comment on `TestSafeResourceName` function) | Update the comment |

---

## DRY Compliance

| Severity | Location | Finding | Recommendation |
|----------|----------|---------|----------------|
| **Medium** | `azblobsas/sas.go:240` + `s3presign/presign.go:369` | `hmacSHA256` function is **character-for-character identical** in both packages | Extract to `pkg/common/crypto.go` |
| **Medium** | `core/backup.go:82-91` + `core/restore.go:90-98` | DPA CRD detection block is **verbatim duplicated** (8 identical lines) | Extract to helper like `common.DetectDPA(ctx, client, log) bool` |
| Medium | `s3presign/presign.go:94-149` + `presign.go:233-270` | INI profile iteration logic (~20 lines) duplicated within the same file between `ParseAWSCredentials` and `parseSTSProfile` | Extract `parseINIProfile(data, profile) map[string]string` |
| Low | `azblobsas/sas.go` + `delegation.go` | SAS URL generation structural overlap (time computation, host/scheme, final URL assembly) | Acceptable — keeping variants self-contained improves security auditability |
| Low | `azblobsas/sas.go:18` + `s3presign/presign.go:19` | `var nowFunc = time.Now` in both packages | Acceptable — standard Go testing pattern, each package needs its own |

### Acceptable non-findings

- `parseDotenv` (azblobsas) vs `ParseAWSCredentials` (s3presign): structurally similar but semantically different (dotenv vs INI-with-sections). Separate implementations justified.
- `uriEncode` in both packages: different encoding behavior (PathEscape vs QueryEscape+SigV4 fixups). Intentional per Azure/AWS signing specs.
- BSL fetch logic in `restore.go` vs `orchestrator.go`: different responsibilities and return types. Too trivial to extract.

---

## SOLID Compliance

| Principle | Verdict | Notes |
|-----------|---------|-------|
| **SRP** | Mostly compliant | `azblobsas` well-decomposed by file. Minor: `s3presign/presign.go` mixes credential parsing + URL generation in one file. |
| **OCP** | Compliant | `signSnapshotURL` extends behavior (Azure branch) without modifying S3 path. `ParseAWSCredentialData` extends credential detection without modifying `ParseAWSCredentials`. |
| **LSP** | Fully compliant | `TokenProvider` and `STSAssumeRoler` implementations (production + mocks) honor contracts. |
| **ISP** | Fully compliant | Both interfaces are single-method and maximally focused. |
| **DIP** | Mostly compliant | `newTokenProvider` and `newSTSClient` function fields on `RestorePlugin` are textbook DIP. Minor: direct `os.Getenv` calls in `presign.go:167` and `restore.go:390-397` bypass injection. |

### SOLID details

**SRP findings:**
- `pkg/azblobsas/` is well-decomposed: `delegation.go` (user delegation key), `sas.go` (Service SAS + credential parsing), `token.go` (AAD token acquisition).
- `pkg/s3presign/sts.go` has one job: STS AssumeRoleWithWebIdentity. Clean separation from presigning.
- `pkg/common/credentials.go` does exactly one thing: resolve a BSL credential reference with fallback.
- Minor: `s3presign/presign.go` mixes credential parsing/detection with SigV4 URL generation.

**DIP findings:**
- `RestorePlugin` depends on abstractions via function fields (`newTokenProvider`, `newSTSClient`) — wired with production implementations in `NewRestorePlugin` and overridden in tests.
- `var httpDo` in `delegation.go:16` is a weaker DIP form compared to `STSClient.HTTPClient` struct field in `sts.go:29`.
- `os.Getenv` calls in `presign.go:167` and `restore.go:390-397` are hard dependencies on the OS environment.

---

## Overall Verdict: PASS WITH RECOMMENDATIONS

This is a well-structured, well-tested PR that adds significant multi-cloud credential support. The architecture decisions (interfaces for testability, dependency injection via function fields, pure-stdlib STS client) are sound. Build, vet, and all tests pass.

### Required Actions

None — no blocking issues.

### Recommended Improvements

1. **`presignAzBlobUrl` → `presignAzBlobURL`** (Go acronym convention) — `restore.go:339,381`
2. **Add `pkg/common/credentials_test.go`** for `ResolveCredentialRef` — two branches, used in critical paths
3. **Extract duplicated `hmacSHA256`** to `pkg/common/crypto.go` — identical function in two packages
4. **Extract DPA detection helper** — 8 verbatim-identical lines in `backup.go` and `restore.go`
5. **Extract `parseINIProfile`** in `s3presign/presign.go` — deduplicate ~20 lines of INI iteration

File: pkg/azblobsas/delegation.go (hunk 1 above)
Comment (jparrill) [reply pkg/azblobsas/delegation.go:16]: Good catch — it is only called once, but the `var` is intentional: `delegation_test.go` replaces `httpDo` with a fake to avoid real HTTP calls (see `TestGetUserDelegationKey`). It's the same pattern we use with `var nowFunc = time.Now` in the SAS package. I can add a short comment above the var if that helps clarify the intent.

File: pkg/azblobsas/delegation.go (hunk 3 above)
Comment (jparrill) [reply pkg/azblobsas/delegation.go:54]: That's Go's [reference time](https://pkg.go.dev/time#pkg-constants) — it's how `time.Format` works in Go. The magic date `2006-01-02T15:04:05Z` is the layout string (Mon Jan 2 15:04:05 MST 2006), not an actual hardcoded date. The output at runtime is the real current timestamp formatted as ISO 8601.

File: pkg/azblobsas/delegation.go (hunk 2 above)
Comment (jparrill) [reply pkg/azblobsas/delegation.go:124]: Answered above — same Go reference time layout.

File: pkg/core/restore.go (hunk 4 above)
Comment (jparrill) [reply pkg/core/restore.go:216]: This is related to this PR — it was added in commit 0537002. When using the `etcdSnapshot` restore method, Velero would restore the etcd StatefulSet with the old (expired) snapshot URL baked into its spec. That conflicts with the fresh pre-signed URL we inject into the HostedControlPlane. Skipping the StatefulSet restore lets the HCP controller reconcile it with the correct URL. I can add a comment in the code explaining the why if you think it'd help.

Comment (mgencur) [reply pkg/core/restore.go:216]: Thanks. It's fine.

File: pkg/azblobsas/delegation.go (hunk 1 above)
Comment (mgencur) [reply pkg/azblobsas/delegation.go:16]: No action needed.
